### PR TITLE
fix swap claim bug for multi-fees

### PR DESCRIPTION
**Pull request by TalDerei**
references https://github.com/penumbra-zone/web/issues/1407

### CODE REVIEW

File: packages/services/src/view-service/fees.ts [hunk 1]
```diff
@@ -1,15 +1,27 @@
 import { AssetId } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
 import { TransactionPlannerRequest } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1/view_pb';
+import { IndexedDbInterface } from '@penumbra-zone/types/indexed-db';
 
-// Attempts to extract a fee token from the assets used in the actions of the planner request
-// Priority in descending order
```
Comment (grod220) [inline packages/services/src/view-service/fees.ts:5]: Why are we removing the code comment?

File: packages/services/src/view-service/fees.test.ts [hunk 2]
```diff
@@ -88,57 +87,7 @@ describe('extractAltFee', () => {
...
-    const result = extractAltFee(request);
-    expect(result.inner).toEqual(auctionWithdrawAuctiontId.inner);
-  });
-
-  it('throws an error when no asset ID is found', () => {
-    const request = new TransactionPlannerRequest({});
-
-    expect(() => extractAltFee(request)).toThrow(
-      'Could not extract alternative fee assetId from TransactionPlannerRequest',
-    );
```
Comment (grod220) [inline packages/services/src/view-service/fees.test.ts:142]: Were these deleted by mistake?

File: packages/services/src/view-service/fees.ts [hunk 3]
```diff
@@ -1,15 +1,27 @@
...
+): Promise<AssetId> => {
   const outputAsset = request.outputs.map(o => o.value?.assetId).find(Boolean);
   if (outputAsset) return outputAsset;
 
   const swapAsset = request.swaps.map(assetIn => assetIn.value?.assetId).find(Boolean);
   if (swapAsset) return swapAsset;
 
+  const swapCommitment = request.swapClaims
+    .map(swapClaim => swapClaim.swapCommitment)
+    .find(commitment => commitment !== undefined);
```
Comment (grod220) [inline packages/services/src/view-service/fees.ts:17]: You can simplify this like the others to:

```diff
- .find(commitment => commitment !== undefined);
+ .find(Boolean)
```

File: packages/services/src/view-service/fees.test.ts
Comment (grod220) [inline packages/services/src/view-service/fees.test.ts:1]: Can you write a test for swap claims as well?

File: packages/services/src/view-service/fees.ts [hunk 4]
```diff
@@ -1,15 +1,27 @@
...
+  if (swapCommitment) {
+    const swaps = await indexedDb.getSwapByCommitment(swapCommitment);
+    if (swaps?.swap?.tradingPair?.asset1) {
+      return swaps.swap.tradingPair.asset1;
+    }
+  }
```
Comment (grod220) [inline packages/services/src/view-service/fees.ts:23]: I believe this is saying "If a user is making a swap claim, attempt to find the assetIn they originally swapped with for an alt fee".

But given swap claims happen in different blocks than swaps, isn't it possible the user doesn't have that original asset anymore? Like what if they swapped the entire balance of it out? Is this logic we can't really depend on?

File: packages/services/src/view-service/transaction-planner/index.ts [hunk 5]
```diff
@@ -14,11 +14,13 @@ export const transactionPlanner: Impl['transactionPlanner'] = async (req, ctx) =
   const { indexedDb } = await services.getWalletServices();
 
   // Query IndexedDB directly to check for the existence of staking token
-  const nativeToken = await indexedDb.hasStakingAssetBalance(req.source);
+  const nativeToken = await indexedDb.hasStakingAssetBalance();
```
Comment (grod220) [inline packages/services/src/view-service/transaction-planner/index.ts:17]: What's the thinking around removing source here? Is the expectation that paying fees should not be scoped to one account and can be paid by any of your accounts? Is this consistent with the core repo? At the moment, in minifront, we've kinda been scoping all activity to a single account in a tx.

File: packages/services/src/view-service/fees.test.ts (hunk 2 above)
Comment (TalDerei) [reply packages/services/src/view-service/fees.test.ts:142]: will fold them back in when we implement auctions seperately in pr. by passing in the idb handle to `extractAltFees`, we're changing the method behavior so it's expected to necessarily change the tests

File: packages/services/src/view-service/fees.ts (hunk 3 above)
Comment (TalDerei) [reply packages/services/src/view-service/fees.ts:17]: simplified

File: packages/services/src/view-service/fees.ts (hunk 4 above)
Comment (TalDerei) [reply packages/services/src/view-service/fees.ts:23]: as discussed, the original asset value balances are irrelevant since the swap pre-pays the swap claim. The assetId however must be consistent between the swap / swap claim and can be retrieved by querying the swap table in the database or by calling the `SwapByCommitment` service implementation using the swap commitment as the key.

**Comment by TalDerei** [comment]
edge case discovered during pairing:
  
1. Swap GM --> GN
2. acquire UM
3. swap claim never hits extractAltFee

screenshot_2024-07-05_at_4 53 17___pm

File: packages/services/src/view-service/fees.test.ts [hunk 6]
```diff
@@ -21,8 +20,8 @@ describe('extractAltFee', () => {
         }),
       ],
     });
-    const result = extractAltFee(request);
-    expect(result.equals(umAssetId)).toBeTruthy();
+    const outputAsset = request.outputs.map(o => o.value?.assetId).find(Boolean);
+    expect(outputAsset!.equals(umAssetId)).toBeTruthy();
```
Comment (jessepinho) [inline packages/services/src/view-service/fees.test.ts:24]: 🚨 This (and the three changed tests below) are no longer testing `extractAltFee()`, since they don't call it. They should be changed to actually call `extractAltFee()`, or else they're essentially useless tests.

Comment (TalDerei) [reply packages/services/src/view-service/fees.test.ts:24]: we're passing in an indexDB now, not sure how we'd simulate this in the test since MockIndexDB has different interface?

**Comment by TalDerei** [comment]
[swapClaims](https://github.com/penumbra-zone/web/blob/swap-claim-bug/packages/services/src/view-service/fees.ts#L15) inside `extractAltFee` now properly retrieve the asset id from the claim fee rather than the pair inputs. This solves a host of canonical ordering bugs.

File: packages/services/src/view-service/fees.test.ts
Comment (TalDerei) [reply packages/services/src/view-service/fees.test.ts:1]: logic has changed with the inclusion of indexDB, so we should change these tests accordingly.

File: packages/services/src/view-service/transaction-planner/index.ts (hunk 5 above)
Comment (Valentine1898) [reply packages/services/src/view-service/transaction-planner/index.ts:18]: I think the issue is that `source` should always be set, and we should not create transaction plans that are not linked to any account (address index)

Comment (Valentine1898) [reply packages/services/src/view-service/transaction-planner/index.ts:18]: The fact that `hasStakingAssetBalance()` now doesn't take AddressIndex into account leads to a bug where the planner thinks he has a native token for the commission but can't spend it. This bug will be reproduced quite often

**Review by grod220 — APPROVED** [review]

File: packages/services/src/view-service/fees.test.ts (hunk 6 above)
Comment (jessepinho) [reply packages/services/src/view-service/fees.test.ts:25]: Un-resolving this because this is a pretty important issue: the tests here are literally not testing anything. It's the equivalent of writing an `add()` function, then writing a test like this:

```ts
it('adds', () => {
  const result = 2 + 2;
  expect(result).toBe(4);
})
```

In this test, we haven't called the `add()` function, so if we break the `add()` function later, the tests won't catch it.

Happy to pair on updating `MockedIndexedDB` if that would help!

**Review by Valentine1898 — APPROVED** [review]
LGTM!

Just one suggestion, I think planner should throw an error if the client has not passed `source` to `TransactionPlannerRequest`

File: packages/services/src/view-service/fees.test.ts (hunk 6 above)
Comment (grod220) [reply packages/services/src/view-service/fees.test.ts:25]: Talked offline. Given this blocks deployment, we want to move forward. However, we should absolutely follow up with this with a separate PR.

**Comment by grod220** [comment]
Pairing together w/ @TalDerei, merging!
